File: cursos/pria/src/csharp/6_Clases.cs
using System;
using Clase;

namespace Clase
{
    public class Punto
    {
        private int x;
        private int y;

        public Punto(int x, int y)
        {
            this.x = x;
            this.y = y;
        }

        public int GetX()
        {
            return x;
        }

        public void SetX(int x)
        {
            this.x = x;
        }

        public int GetY()
        {
            return y;
        }

        public void SetY(int y)
        {
            this.y = y;
        }
    }


    public class Personaje
    {
        public string Nombre { get; private set; }
        public int X { get; private set; } = 50;
        public int SaludMaxima { get; private set; }
        public int Salud { get; private set; }
        public int Velocidad { get; } = 10;
        public int Ataque { get; } = 8;
        public int Defensa { get; } = 4;

        public Personaje(string nombre, int saludMaxima)
        {
            Nombre = nombre;
            SaludMaxima = saludMaxima;
            Salud = saludMaxima;
        }

        public void MoverDerecha()
        {
            X += Velocidad;
        }

        public void RecibirDaño(int daño)
        {
            Salud -= daño;
        }
    }

    public abstract class Animal
    {
        public string Nombre { get; protected set; }
        public int Energia { get; protected set; }

        protected Animal(string nombre, int energia)
        {
            Nombre = nombre;
            Energia = energia;
        }

        public void Comer()
        {
            Energia += 5;
        }

        public void Mover()
        {
            Energia -= 5;
        }
    }

    public class Perro : Animal
    {
        public string Sonido { get; private set; }
        public string Raza { get; private set; }

        public Perro(string nombre, int energia, string sonido, string raza)
            : base(nombre, energia)
        {
            Sonido = sonido;
            Raza = raza;
        }
    }

    public class Gato : Animal
    {
        private string sonido = "¡Miau!";

        // Propiedad de solo lectura que devuelve el valor del campo 'sonido'.
        public string Sonido => sonido;

        public Gato(string nombre, int energia) : base(nombre, energia)
        {
        }

        public void Ronronear()
        {
            sonido = "Prrrr...";
        }
    }
}

// El uso de 'using Clase;' nos ahorra tener que escribir Clase.Punto, Clase.Personaje, etc.
class Program
{
    static void Main()
    {
        // Ejemplo con la clase Punto.
        Punto punto = new Punto(2, 3);
        Clase.Punto punto2 = new Clase.Punto(2, 3);
        int a = punto.GetX();
        int b = punto.GetY();
        Console.WriteLine($"Punto X: {a}, Punto Y: {b}");

        punto.SetX(5);
        punto.SetY(-4);
        Console.WriteLine($"Nuevo Punto X: {punto.GetX()}, Punto Y: {punto.GetY()}");

        // Ejemplo con la clase Personaje.
        Personaje heroe = new Personaje("Reinwald", 50);
        Personaje villano = new Personaje("Zarosh", 40);

        heroe.MoverDerecha();
        Console.WriteLine($"La posición en X del héroe {heroe.Nombre} es {heroe.X}.");

        villano.RecibirDaño(heroe.Ataque);
        Console.WriteLine($"Los puntos de vida de {villano.Nombre} son {villano.Salud}.");

        // Ejemplo con la clase Animal y las subclases Perro y Gato.
        Gato ruperta = new Gato("Ruperta", 100);
        Gato chloe = new Gato("Chloe", 75);

        Console.WriteLine($"La energía de {ruperta.Nombre} es {ruperta.Energia} y su sonido es {ruperta.Sonido}");
        Console.WriteLine($"La energía de {chloe.Nombre} es {chloe.Energia} y su sonido es {chloe.Sonido}");

        chloe.Mover();
        chloe.Comer();
        chloe.Mover();

        ruperta.Ronronear();

        Console.WriteLine($"Ahora, la energía de {chloe.Nombre} es {chloe.Energia}");
        Console.WriteLine($"Ahora, el sonido de {ruperta.Nombre} es {ruperta.Sonido}");

        Perro toby = new Perro("Toby", 50, "¡Woof!", "Golden Retriever");
        Perro akira = new Perro("Akira", 125, "¡Woof, woof!", "Siberian Husky");

        Console.WriteLine($"{toby.Nombre} es un {toby.Raza} y su sonido es {toby.Sonido}");
        Console.WriteLine($"{akira.Nombre} es un {akira.Raza} y su sonido es {akira.Sonido}");
    }
}

File: cursos/pria/src/python/6_clases.py
# Clase Punto
class Punto:
    def __init__(self, x, y):
        # Método constructor que inicializa los atributos 'x' e 'y'
        self._x = x
        self._y = y

    # Propiedad para 'x'
    @property
    def x(self):
        """Obtiene el valor de 'x'."""
        return self._x

    @x.setter
    def x(self, valor):
        """Establece el valor de 'x'."""
        self._x = valor

    # Propiedad para 'y'
    @property
    def y(self):
        """Obtiene el valor de 'y'."""
        return self._y

    @y.setter
    def y(self, valor):
        """Establece el valor de 'y'."""
        self._y = valor


# Clase Personaje
class Personaje:
    def __init__(self, nombre, salud_maxima):
        # Método constructor que inicializa los atributos del personaje
        self._nombre = nombre
        self._x = 50
        self._salud_maxima = salud_maxima
        self._salud = salud_maxima
        self._velocidad = 10
        self._ataque = 8
        self._defensa = 4

    # Propiedades de la clase
    @property
    def nombre(self):
        return self._nombre

    @property
    def x(self):
        return self._x

    @property
    def salud_maxima(self):
        return self._salud_maxima

    @property
    def salud(self):
        return self._salud

    @property
    def velocidad(self):
        return self._velocidad

    @property
    def ataque(self):
        return self._ataque

    @property
    def defensa(self):
        return self._defensa

    # Métodos de la clase
    def mover_derecha(self):
        # Incrementa la posición en X del personaje
        self._x += self._velocidad

    def recibir_daño(self, daño):
        # Reduce la salud del personaje según el daño recibido
        self._salud -= daño


# Clase Animal (clase base)
class Animal:
    def __init__(self, nombre, energia):
        # Método constructor que inicializa 'nombre' y 'energia'
        self._nombre = nombre
        self._energia = energia

    @property
    def nombre(self):
        return self._nombre

    @property
    def energia(self):
        return self._energia

    def comer(self):
        # Incrementa la energía del animal
        self._energia += 5

    def mover(self):
        # Decrementa la energía del animal
        self._energia -= 5


# Clase Perro (hereda de Animal)
class Perro(Animal):
    def __init__(self, nombre, energia, sonido, raza):
        # Llama al constructor de la clase padre 'Animal' usando 'super()'
        super().__init__(nombre, energia)
        self._sonido = sonido
        self._raza = raza

    @property
    def sonido(self):
        return self._sonido

    @property
    def raza(self):
        return self._raza


# Clase Gato (hereda de Animal)
class Gato(Animal):
    def __init__(self, nombre, energia):
        # Llama al constructor de la clase padre 'Animal' usando 'super()'
        super().__init__(nombre, energia)
        self._sonido = "¡Miau!"

    @property
    def sonido(self):
        return self._sonido

    def ronronear(self):
        # Cambia el sonido del gato a 'Prrrr...'
        self._sonido = "Prrrr..."


# Función principal
def main():
    # Ejemplo con la clase Punto
    punto = Punto(2, 3)
    a = punto.x  # Obtiene el valor de 'x' usando la propiedad
    b = punto.y  # Obtiene el valor de 'y' usando la propiedad
    print(f"Punto X: {a}, Punto Y: {b}")

    punto.x = 5  # Establece un nuevo valor para 'x'
    punto.y = -4  # Establece un nuevo valor para 'y'
    print(f"Nuevo Punto X: {punto.x}, Punto Y: {punto.y}")

    # Ejemplo con la clase Personaje
    heroe = Personaje("Reinwald", 50)
    villano = Personaje("Zarosh", 40)

    heroe.mover_derecha()  # El héroe se mueve hacia la derecha
    print(f"La posición en X del héroe {heroe.nombre} es {heroe.x}.")

    villano.recibir_daño(heroe.ataque)  # El villano recibe daño del héroe
    print(f"Los puntos de vida de {villano.nombre} son {villano.salud}.")

    # Ejemplo con la clase Animal y sus subclases
    ruperta = Gato("Ruperta", 100)
    chloe = Gato("Chloe", 75)

    print(f"La energía de {ruperta.nombre} es {ruperta.energia} y su sonido es {ruperta.sonido}")
    print(f"La energía de {chloe.nombre} es {chloe.energia} y su sonido es {chloe.sonido}")

    chloe.mover()  # Chloe se mueve, pierde energía
    chloe.comer()  # Chloe come, gana energía
    chloe.mover()  # Chloe se mueve nuevamente

    ruperta.ronronear()  # Ruperta cambia su sonido a 'Prrrr...'

    print(f"Ahora, la energía de {chloe.nombre} es {chloe.energia}")
    print(f"Ahora, el sonido de {ruperta.nombre} es {ruperta.sonido}")

    toby = Perro("Toby", 50, "¡Woof!", "Golden Retriever")
    akira = Perro("Akira", 125, "¡Woof, woof!", "Siberian Husky")

    print(f"{toby.nombre} es un {toby.raza} y su sonido es {toby.sonido}")
    print(f"{akira.nombre} es un {akira.raza} y su sonido es {akira.sonido}")


if __name__ == "__main__":
    main()
